drivers/usb/device/usb_dc_numaker.c

```diff
@@ -0,0 +1,1901 @@
+	ep_cur->write_fifo_pos += *size_p;
+	ep_cur->write_fifo_free -= *size_p;
+	if (ep_cur->write_fifo_free == 0) {
+		ep_cur->write_fifo_pos = ep_cur->dmabuf_base;
+	}
+
+	return 0;
+}
+
+/* Update EP read/write FIFO on DATA OUT/IN completed */
+static void numaker_usbd_ep_fifo_update(struct numaker_usbd_ep *ep_cur)
+{
+	__ASSERT_NO_MSG(ep_cur->ep_addr_valid);
+	__ASSERT_NO_MSG(ep_cur->dmabuf_valid);
+
+	if (USB_EP_DIR_IS_OUT(ep_cur->ep_addr)) {
+		/* Read FIFO */
+		/* NOTE: For one-shot implementation, FIFO gets updated from empty. */
+		ep_cur->read_fifo_pos = ep_cur->dmabuf_base;
+		/* NOTE: See comment on mxpld_ctrlout for why make one copy of CTRL OUT's MXPLD */
+		if (USB_EP_GET_IDX(ep_cur->ep_addr) == 0) {
+			ep_cur->read_fifo_used = ep_cur->mxpld_ctrlout;
+		} else {
+			ep_cur->read_fifo_used = USBD_GET_PAYLOAD_LEN(ep_cur->usbd_hw_ep_hndl);
+		}
+	} else {
+		/* Write FIFO */
+		/* NOTE: For one-shot implementation, FIFO gets to empty. */
+		ep_cur->write_fifo_pos = ep_cur->dmabuf_base;
+		ep_cur->write_fifo_free = numaker_usbd_ep_fifo_max(ep_cur);
+	}
+}
+
+/* Configure EP DMA buffer */
+static void numaker_usbd_ep_config_dmabuf(struct numaker_usbd_ep *ep_cur, uint32_t dmabuf_base,
+					  uint32_t dmabuf_size)
+{
+	USBD_SET_EP_BUF_ADDR(ep_cur->usbd_hw_ep_hndl, dmabuf_base);
+
+	ep_cur->dmabuf_valid = true;
+	ep_cur->dmabuf_base = dmabuf_base;
+	ep_cur->dmabuf_size = dmabuf_size;
+
+	/* FIFO is implemented based on DMA buffer, which needs reset accordingly. */
+	numaker_usbd_ep_fifo_reset(ep_cur);
+}
+
+/* Abort EP */
+static void numaker_usbd_ep_abort(struct numaker_usbd_ep *ep_cur)
+{
+	struct usb_dc_numaker_device *dev = ep_cur->dev;
+	const struct usb_dc_numaker_config *config = dev->config;
+
+	/* Abort EP on-going transaction */
+	USBD_STOP_TRANSACTION(ep_cur->usbd_hw_ep_hndl);
+
+#if defined(CONFIG_USB_DC_NUMAKER_WORKAROUND_DISALLOW_ISO_IN_OUT_SAME_NUM)
+	if (ep_cur->usbd_hw_ep_cfg & USBD_CFG_TYPE_ISO) {
+		/* Restore to not enabling the function */
+		*((volatile uint32_t *)(config->usbd_base->RESERVE0 + 0)) |= 0x2;
+	}
+#endif
+
+	/* Need to clear NAK for next transaction */
+	ep_cur->nak_clr = false;
+}
+
+/* Configure EP majorly */
+static void numaker_usbd_ep_config_major(struct numaker_usbd_ep *ep_cur,
+					 const struct usb_dc_ep_cfg_data *const ep_cfg)
+{
+	ep_cur->ep_mps_valid = true;
+	ep_cur->ep_mps = ep_cfg->ep_mps;
+
+	/* Configure EP transfer type, DATA0/1 toggle, direction, number, etc. */
+	ep_cur->usbd_hw_ep_cfg =
+		((ep_cfg->ep_type == /* Clear STALL Response in Setup stage */
+		  USB_DC_EP_CONTROL)
+			 ? USBD_CFG_CSTALL
+			 : 0) |
+		(!USBD_CFG_DSQSYNC_Msk) | /* Default to DATA0 */
+		USBD_CFG_EPMODE_DISABLE | /* Endpoint IN/OUT, though, default to disabled */
+		((ep_cfg->ep_type ==      /* Isochronous or not */
+		  USB_DC_EP_ISOCHRONOUS)
+			 ? USBD_CFG_TYPE_ISO
+			 : 0) |
+		((USB_EP_GET_IDX(ep_cfg->ep_addr) /* Endpoint index */
+		  << USBD_CFG_EPNUM_Pos) &
+		 USBD_CFG_EPNUM_Msk);
```

Comment:
this could use some refactoring :-) maybe drop the ternaries, how about something like:

```c
ep_cur->usbd_hw_ep_cfg = 0;

/* Clear STALL Response in Setup stage */
if ((ep_cfg->ep_type == USB_DC_EP_CONTROL) {
        ep_cur->usbd_hw_ep_cfg |= USBD_CFG_CSTALL;
}

/* Default to DATA0 */
ep_cur->usbd_hw_ep_cfg |= !USBD_CFG_DSQSYNC_Msk;

/* Endpoint IN/OUT, though, default to disabled */
ep_cur->usbd_hw_ep_cfg |= USBD_CFG_EPMODE_DISABLEl

/* Isochronous or not */
if (ep_cfg->ep_type == USB_DC_EP_ISOCHRONOUS) {
        ep_cur->usbd_hw_ep_cfg |= USBD_CFG_TYPE_ISO;
}

/* Endpoint index */
ep_cur->usbd_hw_ep_cfg |= (USB_EP_GET_IDX(ep_cfg->ep_addr) << USBD_CFG_EPNUM_Pos) & USBD_CFG_EPNUM_Msk;
```

...also that `!USBD_CFG_DSQSYNC_Msk` does not look quite right

Reply:
still relevant

Reply:
updated. besides, `!USBD_CFG_DSQSYNC_Msk` is changed to:
```c
/* Default to DATA0 */
ep_cur->usbd_hw_ep_cfg &= ~USBD_CFG_DSQSYNC_Msk;
```